contracts/koh/contracts/KingAutomaton.sol

```diff
@@ -324,66 +301,247 @@ contract KingAutomaton {
+    payForGas(id, 1);
+  }
+
+  function createProposal(address payable contributor, string calldata title, string calldata documents_link,
+      bytes calldata documents_hash, uint256 budget_period_len, uint256 num_periods, uint256 budget_per_period)
+      external returns (uint256 _id) {
+    require(num_periods * budget_per_period <= proposals_data.treasuryLimitPercentage * balances[treasuryAddress] / 100);
+    _id = createBallotBox(2);
+    proposals_data.createProposal(_id, contributor, title, documents_link, documents_hash, budget_period_len, num_periods, budget_per_period);
+    transferInternal(treasuryAddress, address(_id), num_periods * budget_per_period);
+  }
+
+  function unpaidSlots(uint256 _id) public view returns (uint256) {
+    Proposals.BallotBox memory b = proposals_data.ballotBoxes[_id];
+    require(b.state != Proposals.BallotBoxState.Uninitialized, "Invalid ballot box ID!");
+    return numSlots - b.paidSlots;
+  }
+
+  // Pay for multiple slots at once, 32 seems to be a reasonable amount.
+  function payForGas(uint256 _id, uint256 _slotsToPay) public {
+    proposals_data.payForGas(_id, _slotsToPay, numSlots);
+  }
+
+  function calcVoteDifference(uint256 _id) view public returns (int256) {
+    return proposals_data.calcVoteDifference(_id, numSlots);
+  }
+
+  function castVote(uint256 _id, uint256 _slot, uint8 _choice) public slotOwner(_slot) {
+    updateProposalState(_id);
+    proposals_data.castVote(_id, _slot, _choice);
+  }
+
+  function getVote(uint256 _id, uint256 _slot) public view returns (uint) {
+    Proposals.BallotBox storage b = proposals_data.ballotBoxes[_id];
+    require(b.state != Proposals.BallotBoxState.Uninitialized, "Invalid ballot box ID!");
+    uint256 bitsPerVote = Helpers.msb(b.numChoices) + 1;
+    uint256 votesPerWord = 255 / bitsPerVote;
+    uint256 index = _slot / votesPerWord;
+    uint256 offset = (_slot % votesPerWord) * bitsPerVote;
+    uint256 mask = ((1 << bitsPerVote) - 1) << offset;
+    return (b.votes[index] & mask) >> offset;
+  }
+
+  function getVoteCount(uint256 _id, uint256 _choice) public view validBallotBoxID(_id) returns(uint256) {
+    return proposals_data.ballotBoxes[_id].voteCount[_choice] & ALL_BUT_MSB;
   }
 
   function completeProposal(uint256 _id) private {
     updateProposalState(_id);
-    BallotBox storage b = ballotBoxes[_id];
-    require(b.state == BallotBoxState.Active, "Ballot is not active!");
-    Proposal storage p = proposals[_id];
-    ProposalState p_state = p.state;
-    require(p_state == ProposalState.Started || p_state == ProposalState.Contested, "Invalid proposal state!");
-    p.state = ProposalState.Completed;
-    b.state = BallotBoxState.Inactive;
+    proposals_data.completeProposal(_id);
   }
 
   function updateProposalState(uint256 _id) public validBallotBoxID(_id) {
-    Proposal storage p = proposals[_id];
-    ProposalState p_state = p.state;
+    Proposals.Proposal storage p = proposals_data.proposals[_id];
```

Comment:
Why wouldn't updateProposalState be part of the Proposals library?
Or at least a portion of this logic seems like should live inside the Proposals library.
(since you probably can't call transferInternal from the library itself)